File: 2-add-two-numbers/2-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) 
    {
        int carry = 0;
        ListNode* temp = new ListNode(0);
        ListNode* head = temp;
        while(l1!=NULL || l2!=NULL)
        {
            int p1 = l1 == NULL ? 0 : l1->val;
            int p2 = l2 == NULL ? 0 : l2->val;
            int val = (p1 + p2 + carry) % 10;
            carry = (int)(p1 + p2 + carry) / 10;
            temp->next = new ListNode(val);
            temp = temp->next;
            l1 = l1 == NULL ? NULL : l1->next;
            l2 = l2 == NULL ? NULL : l2->next;
        }
        if(carry)
        {
            temp->next = new ListNode(1);
        }
        
        return head->next;
    }
};

File: 2-add-two-numbers/2-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        
        dummy = head = ListNode(-1)
        
        c = 0
        
        while l1 or l2:
            
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            
            c, s = divmod((val1 + val2 + c),10)
            
            node = ListNode(s)
            
            head.next = node
            head = head.next
            
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        
        if c:
            node = ListNode(1)
            head.next = node
            
        return dummy.next